File: python/prompt_matching/core/trie.py
"""
Module contenant l'implémentation du Trie.

Le Trie est une structure de données arborescente qui permet de stocker des paires clé-valeur
où les clés sont des séquences (comme des chaînes de caractères), avec des fonctionnalités
de recherche par préfixe.
"""

from typing import TypeVar, Generic, Dict, List, Tuple, Optional, Callable, Iterable, Any

K = TypeVar('K')  # Type de clé (séquence)
E = TypeVar('E')  # Type d'élément de la séquence
V = TypeVar('V')  # Type de valeur

class Trie(Generic[K, E, V]):
    """
    Une structure de données arborescente pour stocker des paires clé-valeur où les clés sont des séquences.
    
    Le Trie permet une recherche efficace par préfixe, ce qui est utile pour la détection
    de signatures de prompts.
    """
    
    class Node:
        """Nœud interne du Trie."""
        
        def __init__(self):
            """Initialise un nouveau nœud."""
            self.children: Dict[E, Trie.Node] = {}
            self.value: Optional[V] = None
            self.has_value: bool = False
    
    def __init__(self, key_to_elements: Callable[[K], Iterable[E]] = None):
        """
        Initialise une nouvelle instance de Trie.
        
        Args:
            key_to_elements: Fonction de conversion d'une clé en séquence d'éléments
                            (si None, la clé doit être itérable)
        """
        self._root = self.Node()
        self._count = 0
        self._key_to_elements = key_to_elements or (lambda k: k)
    
    @property
    def count(self) -> int:
        """Nombre d'éléments dans le Trie."""
        return self._count
    
    def add(self, key: K, value: V) -> None:
        """
        Ajoute ou met à jour une paire clé-valeur dans le Trie.
        
        Args:
            key: La clé à ajouter (séquence)
            value: La valeur associée à la clé
            
        Raises:
            ValueError: Si la clé est None
        """
        if key is None:
            raise ValueError("La clé ne peut pas être None")
        
        elements = self._key_to_elements(key)
        node = self._root
        
        # Parcourir ou créer le chemin pour la clé
        for element in elements:
            if element not in node.children:
                node.children[element] = self.Node()
            node = node.children[element]
        
        # Mettre à jour le compteur si c'est une nouvelle clé
        if not node.has_value:
            self._count += 1
        
        # Stocker la valeur
        node.value = value
        node.has_value = True
    
    def try_get_value(self, key: K) -> Tuple[bool, Optional[V]]:
        """
        Tente de récupérer la valeur associée à une clé.
        
        Args:
            key: La clé à rechercher
            
        Returns:
            Un tuple (succès, valeur) où succès est True si la clé existe,
            et valeur est la valeur associée à la clé ou None si la clé n'existe pas
            
        Raises:
            ValueError: Si la clé est None
        """
        if key is None:
            raise ValueError("La clé ne peut pas être None")
        
        node = self._find_node(key)
        
        if node is not None and node.has_value:
            return True, node.value
        return False, None
    
    def try_get_value_by_prefix(self, prefix: K) -> Tuple[bool, Optional[V]]:
        """
        Tente de récupérer la valeur associée à la plus longue clé qui est un préfixe de la séquence donnée.
        
        Args:
            prefix: La séquence à rechercher
            
        Returns:
            Un tuple (succès, valeur) où succès est True si un préfixe existe,
            et valeur est la valeur associée au préfixe le plus long ou None si aucun préfixe n'existe
            
        Raises:
            ValueError: Si le préfixe est None
        """
        if prefix is None:
            raise ValueError("Le préfixe ne peut pas être None")
        
        elements = list(self._key_to_elements(prefix))
        node = self._root
        last_value_node = None if not node.has_value else node
        
        # Parcourir le Trie en suivant le préfixe
        for element in elements:
            if element not in node.children:
                break
            node = node.children[element]
            if node.has_value:
                last_value_node = node
        
        if last_value_node is not None:
            return True, last_value_node.value
        return False, None
    
    def remove(self, key: K) -> bool:
        """
        Supprime une paire clé-valeur du Trie.
        
        Args:
            key: La clé à supprimer
            
        Returns:
            True si la clé a été supprimée, False si elle n'existait pas
            
        Raises:
            ValueError: Si la clé est None
        """
        if key is None:
            raise ValueError("La clé ne peut pas être None")
        
        elements = list(self._key_to_elements(key))
        return self._remove_recursive(self._root, elements, 0)
    
    def clear(self) -> None:
        """Supprime tous les éléments du Trie."""
        self._root = self.Node()
        self._count = 0
    
    def _find_node(self, key: K) -> Optional[Node]:
        """
        Trouve le nœud correspondant à une clé.
        
        Args:
            key: La clé à rechercher
            
        Returns:
            Le nœud correspondant à la clé ou None si la clé n'existe pas
        """
        elements = self._key_to_elements(key)
        node = self._root
        
        for element in elements:
            if element not in node.children:
                return None
            node = node.children[element]
        
        return node
    
    def _remove_recursive(self, node: Node, elements: List[E], depth: int) -> bool:
        """
        Supprime récursivement une clé du Trie.
        
        Args:
            node: Le nœud courant
            elements: La liste des éléments de la clé
            depth: La profondeur actuelle dans l'arbre
            
        Returns:
            True si la clé a été supprimée, False sinon
        """
        # Si nous avons atteint la fin de la clé
        if depth == len(elements):
            if not node.has_value:
                return False
            
            node.has_value = False
            node.value = None
            self._count -= 1
            return True
        
        element = elements[depth]
        
        if element not in node.children:
            return False
        
        result = self._remove_recursive(node.children[element], elements, depth + 1)
        
        # Si le nœud enfant n'a plus d'enfants et pas de valeur, on peut le supprimer
        child = node.children[element]
        if not child.has_value and not child.children:
            del node.children[element]
        
        return result

File: dotnet/src/Connectors/Connectors.AI.MultiConnector/PromptMatching/Trie.cs
// Copyright (c) MyIA. All rights reserved.

using System;
using System.Collections.Generic;

namespace MyIA.SemanticKernel.Connectors.AI.MultiConnector.PromptMatching
{
    /// <summary>
    /// Implémentation générique d'un arbre à préfixe (Trie)
    /// </summary>
    /// <typeparam name="K">Type de clé (typiquement string)</typeparam>
    /// <typeparam name="C">Type de caractère (typiquement char)</typeparam>
    /// <typeparam name="V">Type de valeur associée</typeparam>
    public class Trie<K, C, V> : ITrie<K, C, V> where K : IEnumerable<C>
    {
        /// <summary>
        /// Nœud interne de l'arbre à préfixe
        /// </summary>
        protected class TrieNode
        {
            /// <summary>
            /// Enfants du nœud, indexés par caractère
            /// </summary>
            public HybridDictionary<C, TrieNode> Children { get; } = new HybridDictionary<C, TrieNode>();

            /// <summary>
            /// Indique si ce nœud est la fin d'une clé
            /// </summary>
            public bool IsEndOfKey { get; set; }

            /// <summary>
            /// Valeur associée à la clé se terminant à ce nœud
            /// </summary>
            public V? Value { get; set; }
        }

        /// <summary>
        /// Racine de l'arbre
        /// </summary>
        protected readonly TrieNode Root = new TrieNode();

        /// <summary>
        /// Nombre d'éléments dans l'arbre
        /// </summary>
        public int Count { get; protected set; }

        /// <summary>
        /// Ajoute ou met à jour une valeur associée à une clé
        /// </summary>
        /// <param name="key">Clé à ajouter</param>
        /// <param name="value">Valeur à associer à la clé</param>
        public virtual void Add(K key, V value)
        {
            if (key == null)
            {
                throw new ArgumentNullException(nameof(key));
            }

            TrieNode current = Root;
            bool isNewKey = false;

            foreach (C c in key)
            {
                if (!current.Children.TryGetValue(c, out TrieNode? child))
                {
                    child = new TrieNode();
                    current.Children.Add(c, child);
                }

                current = child;
            }

            isNewKey = !current.IsEndOfKey;
            current.IsEndOfKey = true;
            current.Value = value;

            if (isNewKey)
            {
                Count++;
            }
        }

        /// <summary>
        /// Recherche une valeur associée à une clé exacte
        /// </summary>
        /// <param name="key">Clé à rechercher</param>
        /// <param name="value">Valeur associée à la clé si trouvée</param>
        /// <returns>True si la clé existe, false sinon</returns>
        public virtual bool TryGetValue(K key, out V value)
        {
            if (key == null)
            {
                throw new ArgumentNullException(nameof(key));
            }

            TrieNode? node = FindNode(key);

            if (node != null && node.IsEndOfKey)
            {
                value = node.Value!;
                return true;
            }

            value = default!;
            return false;
        }

        /// <summary>
        /// Recherche une valeur associée à un préfixe
        /// </summary>
        /// <param name="prefix">Préfixe à rechercher</param>
        /// <param name="value">Valeur associée au préfixe le plus long correspondant</param>
        /// <returns>True si un préfixe correspondant existe, false sinon</returns>
        public virtual bool TryGetValueByPrefix(K prefix, out V value)
        {
            if (prefix == null)
            {
                throw new ArgumentNullException(nameof(prefix));
            }

            TrieNode current = Root;
            TrieNode? lastMatchingNode = null;

            foreach (C c in prefix)
            {
                if (!current.Children.TryGetValue(c, out TrieNode? child))
                {
                    break;
                }

                current = child;

                if (current.IsEndOfKey)
                {
                    lastMatchingNode = current;
                }
            }

            if (lastMatchingNode != null)
            {
                value = lastMatchingNode.Value!;
                return true;
            }

            value = default!;
            return false;
        }

        /// <summary>
        /// Supprime une clé et sa valeur associée
        /// </summary>
        /// <param name="key">Clé à supprimer</param>
        /// <returns>True si la clé a été supprimée, false si elle n'existait pas</returns>
        public virtual bool Remove(K key)
        {
            if (key == null)
            {
                throw new ArgumentNullException(nameof(key));
            }

            bool removed = RemoveRecursive(Root, key.GetEnumerator(), 0);

            if (removed)
            {
                Count--;
            }

            return removed;
        }

        /// <summary>
        /// Efface toutes les entrées de l'arbre
        /// </summary>
        public virtual void Clear()
        {
            Root.Children.Clear();
            Count = 0;
        }

        /// <summary>
        /// Recherche un nœud correspondant à une clé
        /// </summary>
        /// <param name="key">Clé à rechercher</param>
        /// <returns>Nœud correspondant à la clé, ou null si non trouvé</returns>
        protected TrieNode? FindNode(K key)
        {
            TrieNode current = Root;

            foreach (C c in key)
            {
                if (!current.Children.TryGetValue(c, out TrieNode? child))
                {
                    return null;
                }

                current = child;
            }

            return current;
        }

        /// <summary>
        /// Supprime récursivement une clé de l'arbre
        /// </summary>
        /// <param name="node">Nœud courant</param>
        /// <param name="keyEnumerator">Énumérateur de la clé</param>
        /// <param name="depth">Profondeur actuelle dans l'arbre</param>
        /// <returns>True si la clé a été supprimée, false sinon</returns>
        protected bool RemoveRecursive(TrieNode node, IEnumerator<C> keyEnumerator, int depth)
        {
            if (!keyEnumerator.MoveNext())
            {
                // Fin de la clé, vérifier si c'est une clé valide
                if (!node.IsEndOfKey)
                {
                    return false;
                }

                node.IsEndOfKey = false;
                node.Value = default;

                // Si le nœud n'a pas d'enfants, il peut être supprimé
                return node.Children.Count == 0;
            }

            C currentChar = keyEnumerator.Current;

            if (!node.Children.TryGetValue(currentChar, out TrieNode? child))
            {
                return false;
            }

            bool shouldRemoveChild = RemoveRecursive(child, keyEnumerator, depth + 1);

            if (shouldRemoveChild)
            {
                node.Children.Remove(currentChar);

                // Si ce nœud n'est pas la fin d'une clé et n'a pas d'autres enfants, il peut être supprimé
                return !node.IsEndOfKey && node.Children.Count == 0;
            }

            return false;
        }
    }
}
